tools/flow_grid.py

```diff
@@ -0,0 +1,695 @@
+               appropriate value ('ascii' or 'raster').
+        data_type : 'dem', 'dir', 'acc' or other string
+                     Name of dataset. Will determine the name of the attribute
+                     representing the gridded data. Default values are used
+                     internally by some class methods:
+                         'dem' : digital elevation data
+                         'dir' : flow direction data
+                         'acc' : flow accumulation (upstream area) data
+                    
+        input_type : 'raster', 'ascii' or 'array'
+                     Type of input data.
+        band : int
+               For raster data, the band number to read.
+        nodata : int or float
+                 Value indicating no data.
+        bbox : tuple or list
+               Bounding box, if none provided.
+
+        """
+
+        # read ascii file
+        if input_type == 'ascii':
+            with open(data) as header:
+                ncols = int(header.readline().split()[1])
+                nrows = int(header.readline().split()[1])
+                xll = ast.literal_eval(header.readline().split()[1])
+                yll = ast.literal_eval(header.readline().split()[1])
+                cellsize = ast.literal_eval(header.readline().split()[1])
+                nodata = ast.literal_eval(header.readline().split()[1])
+                shape = (nrows, ncols)
+                bbox = (xll, yll, xll + ncols*cellsize, yll + nrows*cellsize)
+            data = np.loadtxt(data, skiprows=6, **kwargs)
+            nodata = data.dtype.type(nodata)
+
+        # read raster file
+        if input_type == 'raster':
+            import rasterio
+            f = rasterio.open(data)
+            crs = f.crs
+            bbox = tuple(f.bounds)
+            shape = f.shape
+            cellsize = f.affine[0]
+            nodata = f.nodatavals[0]
+            if len(f.indexes) > 1:
+                data = np.ma.filled(f.read_band(band))
+            else:
+                data = np.ma.filled(f.read())
+                f.close()
+                data = data.reshape(shape)
+            nodata = data.dtype.type(nodata)
+
+        # read numpy array
+        if input_type == 'array':
+            shape = data.shape
+
+        # if there are no datasets, initialize bbox, shape,
+        # cellsize and crs based on incoming data
+        if len(self.grid_props.keys()) < 1:
+            self.bbox = bbox
+            self.shape = shape
+            self.cellsize = cellsize
+            self.crs = crs
+            self.mask = np.ones(self.shape, dtype=np.bool)
+            self.shape_min = np.min_scalar_type(max(self.shape))
+            self.size_min = np.min_scalar_type(data.size)
+
+        # if there are existing datasets, conform incoming
+        # data to bbox
+        else:
+            try:
+                np.testing.assert_almost_equal(cellsize, self.cellsize)
```

Comment:
do you know what the difference between `numpy.allclose` and `numpy.testing.assert_almost_equal` is?  You use the latter, which is fine, but I generally to avoid using `assert` and raise specific errors instead.